test/xbgpu/test_engine.py

```diff
@@ -1046,28 +1079,41 @@ def heap_factory(batch_index: int, present: np.ndarray) -> list[spead2.send.Heap
+        weights = rng.uniform(0.5, 2.0, size=(N_BEAM_REQUESTS, len(beam_outputs), n_ants))
+        quant_gains = rng.uniform(0.5, 2.0, size=(N_BEAM_REQUESTS, len(beam_outputs)))
+        delays = np.zeros((N_BEAM_REQUESTS, len(beam_outputs), n_ants, 2), np.float64)
         # Delay is in seconds, so needs to be very small
-        delays[..., 0] = rng.uniform(-1e-9, 1e-9, size=(len(beam_outputs), n_ants))
+        delays[..., 0] = rng.uniform(-1e-9, 1e-9, size=(N_BEAM_REQUESTS, len(beam_outputs), n_ants))
         # Phase is in radians
-        delays[..., 1] = rng.uniform(-2 * np.pi, 2 * np.pi, size=(len(beam_outputs), n_ants))
-
-        katcp_requests: dict[str, list[tuple[str, ...]]] = {
-            output.name: [
-                ("beam-weights", output.name, *weights[i]),
-                ("beam-quant-gains", output.name, quant_gains[i]),
-                ("beam-delays", output.name, *[f"{d[0]}:{d[1]}" for d in delays[i]]),
+        delays[..., 1] = rng.uniform(-2 * np.pi, 2 * np.pi, size=(N_BEAM_REQUESTS, len(beam_outputs), n_ants))
+
+        # NOTE: This list of lists is done in an order that has a set of requests for alternating
+        # `beam_outputs`. That is, each item is a standalone set of requests per `beam_output`.
+        # As a result, when issuing these requests, we can use a subset each time we need the
+        # `?beam` parameters to change for all `beam_outputs`.
+        katcp_requests = [
+            [
+                ("beam-weights", boutput.name, *weights[request_id, boutput_id]),
+                ("beam-quant-gains", boutput.name, quant_gains[request_id, boutput_id]),
+                ("beam-delays", boutput.name, *[f"{d[0]}:{d[1]}" for d in delays[request_id, boutput_id]]),
             ]
-            for i, output in enumerate(beam_outputs)
-        }
-        flattened_requests = list(chain.from_iterable(katcp_requests.values()))
+            for request_id in range(N_BEAM_REQUESTS)
+            for boutput_id, boutput in enumerate(beam_outputs)
+        ]
+        # NOTE: This value is arbitrarily chosen, but must be less than
+        # `N_TOTAL_XB_HEAPS` / `HEAPS_PER_FENGINE_PER_CHUNK` to actually take effect. This, as
+        # each call to `patch_get_in_item` grabs a Chunk containing `HEAPS_PER_FENGINE_PER_CHUNK`
+        # heaps.
+        beam_params_change_index = 10
+        # NOTE: We use the 'upper' subset of `katcp_requests` here as `patch_get_in_item` is
+        # triggered during the BPipeline's processing of data. We use the 'lower' subset for
+        # `?beam` requests before it starts processing data so as to make data verification a
+        # bit simpler.
         steady_state_timestamps = self._patch_get_in_item(
             monkeypatch,
-            count=10,
+            count=beam_params_change_index,
             client=client,
-            requests=flattened_requests,
+            requests=list(chain.from_iterable(katcp_requests[N_BEAM_REQUESTS:])),
```

Comment:
This indexing doesn't look right - I think it should be `[len(beam_outputs):]`. But you'll probably make the code clearer if you add another level to katcp_requests, so that it's a list of list of lists of shape (N_BEAM_REQUESTS, len(beam_outputs), 3).